### ISSUE TITLE
rake assets:precompile fails, because yarn install hasn't been executed

### ISSUE THREAD

**Issue by tbuehl** [heroku/heroku-buildpack-ruby#1240]
Hey, 

My deployments (Rails 7.0.0.alpha2) recently stopped working, because `rake assets:precompile` relies on `node_modules` being installed before my assets are precompiled. This worked before without a problem. Unfortunately I don't know which buildpack version though.

Here you can see a shortened log:

```
remote: -----> Cleaning up...        
remote: -----> Building foodlane from herokuish        
remote: -----> Adding BUILD_ENV to build environment...        
remote:        BUILD_ENV added successfully        
remote:        -----> Warning: Multiple default buildpacks reported the ability to handle this app. The first buildpack in the list below will be used.        
remote:               Detected buildpacks: multi ruby nodejs        
remote:        -----> Multipack app detected  
remote: =====> Downloading Buildpack: https://github.com/heroku/heroku-buildpack-ruby        
remote: =====> Detected Framework: Ruby        
remote: -----> Installing bundler 2.2.21       
...
Using webpacker 5.2.1
Bundle complete! 36 Gemfile dependencies, 116 gems now installed.
Gems in the groups 'development' and 'test' were not installed.
Bundled gems are installed into `./vendor/bundle`
Bundle completed (0.84s)
Cleaning up the bundler cache.
Removing bundler (2.2.21)
-----> Installing node-v16.13.1-linux-x64
-----> Installing yarn-v1.22.17
-----> Detecting rake tasks
-----> Preparing app for Rails asset pipeline
Running: rake assets:precompile
rake aborted!
SassC::SyntaxError: Error: File to import not found or unreadable: bootstrap/scss/bootstrap.
on line 2:1 of app/assets/stylesheets/lib/framework.scss
from line 14:2 of app/assets/stylesheets/application.scss
>> @import "bootstrap/scss/bootstrap";
```

As you can see, `node` and `yarn` are installed, but `yarn install` is not called before `rake assets:precompile`. `package.json` and `yarn.lock` are in the repo.

My current workaround is to set the `heroku/nodejs` buildpack before `heroku/ruby`, but it worked before without that.

**Comment by schneems** [comment]
The ruby buildpack does not call `yarn install`. Our contract with applications is that we call `rake assets:precompile`. Yarn is invoked because webpacker hooks into `rake assets:precompile` to execute `rake yarn:install`. 

This looks like an upstream issue with jsbundling (or whatever method you're using for packaging assets).

I believe that Heroku should work out of the box with jsbundling. If it doesn't can you provide me with an example app that reproduces the problem (though note I'll be OOO all next week). https://codetriage.com/example_app

**Comment by Pauloparakleto** [comment]
@schneems and @tbuehl Yesterday the deploy was ok with the same configuration.
Today none of all branches I tested and previous releases could be deployed as well.
As the last try, I built a minimal project whose behavior is the same as my app.
I checked your link instruction, @schneems , [This is the repository](https://github.com/Pauloparakleto/deploy-fails-heroku-rails).

Expected behavior: a minimal rails 6 app must be deployed to heroku. I hope it can help.

**Comment by MoKuH** [comment]
Same issue as @Pauloparakleto 
`       Module build failed (from ./node_modules/sass-loader/dist/cjs.js):
       Error: Node Sass does not yet support your current environment: Linux 64-bit with Unsupported runtime (93)
       For more information on which environments are supported please see:
       https://github.com/sass/node-sass/releases/tag/v4.14.1`
       
       Every thing was ok yesterday. I expect the issue is linked with the Heroku's release related to the yarn and npm versions
       
       https://devcenter.heroku.com/changelog-items/2306

**Comment by Pauloparakleto** [comment]
@MoKuH it is seems we now need to use explicitly our same versions that is running ok on local and tests pipeline environments.
If I have a time I'll try locally these node version settings in a kinda sandbox branch. Let me know if you have tried it today. Thanks.

**Comment by MoKuH** [comment]
@Pauloparakleto I already tried using the 16.13.1 node version locally which is working. But I still have the same issue on Heroku.

**Comment by Pauloparakleto** [comment]
@MoKuH Have you tried precompile locally and then push it? I didn't try it yet. Doing it the command asset precompile from heroku buildpack won't run assuming yours.

**Comment by tbuehl** [comment]
@schneems I can confirm that my problem is unrelated to the buildpack. 

When emptying `node_modules` and running `rails assets:precompile` locally, the errors happens too. 

When I check out an earlier Rails version, `yarn install` is called (like expected) before compiling the assets, and everything works fine. I don't know where the error is yet, but I think we can close this issue.

@Pauloparakleto @MoKuH Your error is unrelated to mine from what I can see. I think it would be better to open a separate issue?

**Comment by tbuehl** [comment]
@schneems BTW I am using the `webpacker (5.4.3)` gem, not `css-bundling` on this project.

**Comment by tbuehl** [comment]
I found the fix: For Rails 7, in my case either `7.0.0.alpha2` or `7.1.0.alpha`, I had to upgrade the `webpacker` gem to `6.0.0.rc.6` to make it work again.

**Comment by Pauloparakleto** [comment]
@tbuehl Let me just try your way so we can close this after. Yesterday I tried locally a rails 6 with the webpacker pointing to the main and the assets precompile works fine. Let me try deploying the minimal project I built following the @schneems instructions. thanks for your answer about webpack.

**Comment by Pauloparakleto** [comment]
Checked with a [rails 6 application](https://github.com/Pauloparakleto/deploy-fails-heroku-rails).
@MoKuH you're right the [change in the node default modules version ](https://devcenter.heroku.com/changelog-items/2306)announced at 16 December 2121 obligated us to set exactly our local  environment to match the heroku machine settings.

And @tbuehl the webpack was indeed the key.

Steps to deploy:
1) I set ruby to a maintained one, 2.7.5.
2) I followed the main branch webpack rails installation.
3) Checked the `rake assets:precompile RAILS_ENV=production` works fine locally.
Obs: If there be an error saying `browserslisterror contains both .browserslistrc and package.json with browsers” `
Just remove the `.browserslistrc`. And try precompile again.
4) Commit and push the changes.
5) Deployed!
